traitsui/tests/editors/test_instance_editor.py

```diff
@@ -85,8 +90,46 @@ def test_simple_editor_resynch_editor(self):
             self.assertEqual(displayed, "bye")
 
     def test_simple_editor_parent_closed(self):
-        obj = NonmodalInstanceEditor()
+        obj = ObjectWithInstance()
         tester = UITester()
         with tester.create_ui(obj, dict(view=get_view('simple'))) as ui:
             instance = tester.find_by_name(ui, "inst")
             instance.perform(MouseClick())
+
+    @unittest.skipIf(no_modal_dialog_tester, "ModalDialogTester unavailable")
+    def test_simple_editor_modal(self):
+        # Test launching the instance editor with kind set to 'modal'
+        obj = ObjectWithInstance()
+        ui_tester = UITester()
+        view = View(
+            Item("inst", style="simple", editor=InstanceEditor(kind="modal"))
+        )
+
+        with ui_tester.create_ui(obj, dict(view=view)) as ui:
+
+            def click_button():
+                ui_tester.find_by_name(ui, "inst").perform(MouseClick())
+
+            def when_opened(tester):
+                with tester.capture_error():
+                    try:
+                        dialog_ui = tester.get_dialog_widget()._ui
```

Comment:
seems like this is the "standard way" to do this?  See my comment on the documentation

Reply:
I am reluctant to mention this in the documentation because the `_ui` reference looks really dodgy (the fact that it is privately named gives away some hints too!). What we have is a cycle reference from UI to a GUI toolkit widget back to the UI. Unfortunately this does seem to be the only way, _right_now_, for getting the UI via the ModalDialogTester.

A lot of times when we write tests for code written without tests, we have to add questionable references and hooks just so that we can add those tests, but then the end goal is that these tests will then allow us to refactor the code safely so that eventually we don't need these hooks any more.

Reply:
More thoughts on how it should have been:
In this block:
https://github.com/enthought/traitsui/blob/1a83c8a3b691ab204cd89e41ed810a2e7d21c3fe/traitsui/qt4/instance_editor.py#L429-L431
The `UI` instance is created within `edit_traits` and if `kind` is modal, the modal dialog is open and it blocks the function from returning until the dialog is closed. So there is no way to get to the UI from there before the dialog is closed. Ideally, the actions for creating an instance of UI and showing the blocking dialog should be separated so that we can keep a reference to the UI before allowing it to show and block. Then we would have been able to use the UITester to get to this dialog UI from the editor, instead of going via the toolkit control.

The reference to `_ui` is assigned here.
https://github.com/enthought/traitsui/blob/1a83c8a3b691ab204cd89e41ed810a2e7d21c3fe/traitsui/qt4/ui_panel.py#L72
https://github.com/enthought/traitsui/blob/1a83c8a3b691ab204cd89e41ed810a2e7d21c3fe/traitsui/wx/ui_panel.py#L100
It is not clear to me yet whose audience this private attribute is intended for, but I would guess it is not a sanctioned hooks for purposes like testing.